pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,210 @@
+	if !found {
+		ipc.identityToIPCache[identity] = map[string]struct{}{}
+	}
+	ipc.identityToIPCache[identity][endpointIP] = struct{}{}
+	endpointIPs := ipc.identityToIPCache[identity]
+
+	// Sort IPs for Upsert into xdsResourceMutator.
+	ipStrings := make([]string, 0, len(endpointIPs))
+	for endpointIP := range endpointIPs {
+		ipStrings = append(ipStrings, endpointIP)
+	}
+	sort.Strings(ipStrings)
+	ipc.xdsResourceMutator.Upsert(envoy.NetworkPolicyHostsTypeURL, identity.StringID(), &envoyAPI.NetworkPolicyHosts{Policy: uint64(identity), HostAddresses: ipStrings}, false)
+
+}
+
+// Delete removes the provided IP-to-security-identity mapping from both caches
+// within ipc.
+func (ipc *IPCache) Delete(endpointIP string) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+	identity, found := ipc.ipToIdentityCache[endpointIP]
+	if found {
+		delete(ipc.ipToIdentityCache, endpointIP)
+		delete(ipc.identityToIPCache[identity], endpointIP)
+		if len(ipc.identityToIPCache[identity]) == 0 {
+			delete(ipc.identityToIPCache, identity)
+		}
+		ipc.xdsResourceMutator.Delete(envoy.NetworkPolicyHostsTypeURL, identity.StringID(), false)
+	}
+}
+
+// LookupByIP returns the corresponding security identity that endpoint IP maps
+// to within the provided IPCache, as well as if the corresponding entry exists
+// in the IPCache.
+func (ipc *IPCache) LookupByIP(endpointIP string) (identity.NumericIdentity, bool) {
+	ipc.mutex.RLock()
+	identity, exists := ipc.ipToIdentityCache[endpointIP]
+	ipc.mutex.RUnlock()
+	return identity, exists
+}
+
+// LookupByIdentity returns the set of endpoint IPs that have security identity
+// ID, as well as if the corresponding entry exists in the IPCache.
+func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (map[string]struct{}, bool) {
+	ipc.mutex.RLock()
+	ips, exists := ipc.identityToIPCache[id]
+	ipc.mutex.RUnlock()
+	return ips, exists
+}
+
+// IPIdentityMappingOwner is the interface the owner of an identity allocator
+// must implement
+type IPIdentityMappingOwner interface {
+	// TriggerPolicyUpdates will be called whenever a policy recalculation
+	// must be triggered
+	TriggerPolicyUpdates(force bool) *sync.WaitGroup
+}
+
+// GetIPIdentityMapModel returns all known endpoint IP to security identity mappings
+// stored in the key-value store.
+func GetIPIdentityMapModel() {
+	// TODO (ianvernon) return model of ip to identity mapping. For use in CLI.
+	// see GH-2555
+}
+
+func ipIdentityWatcher(owner IPIdentityMappingOwner) {
+
+watch:
+	watcher := kvstore.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
+	for {
+		var (
+			identity     identity.NumericIdentity
+			cacheChanged bool
+		)
+
+		// Get events from channel as they come in.
+		event, ok := <-watcher.Events
+
+		// If for whatever reason channel is closed for watcher, try to list and
+		// watch again.
+		if !ok {
+			log.Debugf("%s closed, restarting watch", watcher.String())
+			goto watch
+		}
+
+		_ = json.Unmarshal(event.Value, &identity)
```

Comment:
Shouldn't we handle the error here? even if we could proceed on a partially filled in `identity` object, it would be good to log something for when we start corrupting our cache.

Reply:
Done.